File: SEPTEMBER-2025/16-09-2025/p2197.py
from typing import List
from math import gcd

class Solution:
    def LCM(self, a: int, b: int) -> int:
        return a * b // gcd(a, b)

    def replaceNonCoprimes(self, nums: List[int]) -> List[int]:
        stack = []
        for num in nums:
            while stack and gcd(stack[-1], num) > 1:
                num = self.LCM(stack.pop(), num)
            stack.append(num)  
        return stack

File: SEPTEMBER-2025/16-09-2025/p2197.cpp
class Solution {
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        stack<int> st;
        st.push(nums[0]);
        
        for(int i=1 ; i < nums.size();i++){
             // calculate the st/top
             int b = nums[i];
             if(st.empty()==false && gcd(st.top(),b)<=1){
                st.push(nums[i]);

             }
             else{
                while(st.empty()==false && gcd(st.top(),b)>1){
                    int a = st.top();
                    st.pop();
                    b= lcm(a,b);
                }
                st.push(b);
             }
        }

        vector<int> ans;
        while(st.empty()==false){
            ans.push_back(st.top());
            st.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};